src/main/java/org/apache/sling/resourceresolver/impl/mapping/MapEntries.java

```diff
@@ -1255,6 +1279,33 @@ private boolean loadVanityPath(final Resource resource, final Map<String, List<M
         return hasVanityPath;
     }
 
+    private String createVanityPathQuery() {
+        final String query =
+            VANITY_PATH_BASE_QUERY_DEFAULT +
+                " WHERE sling:vanityPath IS NOT NULL" +
+                Stream
+                    .of(true, false)
+                    .map(this::createVanityPathQueryPathRestriction)
```

Comment:
After reading the code generating the restrictions several times, I believe it is actually correct. But it is not very readable or intuitive.

Maybe it would become more readable if you separated out all "include" restrictions and all "exclude" restrictions, each into their own collection, and then join them with the appropriate "AND" or "OR" operators?

Pseudo code:
```
includesExcludes = Stream.of(new VanityPathConfig(JCR_SYSTEM_PATH, true), factory.getVanityPathConfig().stream())
    .collect(Collectors.groupingBy(VanityPathConfig::isExclude)

excludes = includesExclude.get(Boolean.TRUE)
includes =  includesExclude.get(Boolean.FALSE)

excludesPart =excludes.stream().map("NOT ISDESCENDANTNODE(page, '%s')").join(" AND ")
includesPart = includes.stream().map("ISDESCENDANTNODE(page, '%s')").join(" OR ")

queryRestrictions = String.format("(%s) AND (%s)", excludesPart, includesPart)
```

Beware, I may have gotten the "OR"s and "AND"s wrong ;)

PS: maybe also move `new VanityPathConfig(JCR_SYSTEM_PATH, true)` into `factory.getVanityPathConfig()`?

Reply:
Good points! I will have another look at this. Also, a good point about moving the `/jcr:system`-exclude to the factory.

Reply:
I just had another go at it, @jsedding , does this already look a bit better?